[Api] support for adding default tags to tracer

alimahboubi opened the pull request:
Fixes #6016
Design discussion issue #6016

## Changes

This pull request adds a new overload to the `TracerProvider.GetTracer` method to support the inclusion of tags, similar to the new `ActivitySource` constructor overload in .NET 9. This allows users to associate initial tags with a `Tracer` (and therefore the underlying `ActivitySource`), which can be used to differentiate application instances with different configurations.  These tags are applied to all activities created by the `ActivitySource`.

The new overload is:

```csharp
public Tracer GetTracer(string name, string? version = null, IEnumerable<KeyValuePair<string, object?>>? tags = null)

Kielek commented:
@alimahboubi, it is good practice to execute build locally. In this case it should be done in `Release` mode. For now, I am not able to compile the code.

Could you please fix it?

alimahboubi commented:
@Kielek Thanks for pointing that out! I've fixed the compilation error.  I've also run a local build in Release mode to confirm

Kielek commented on src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Shipped.txt:
```diff
@@ -151,7 +151,9 @@ OpenTelemetry.Trace.Tracer.StartRootSpan(string! name, OpenTelemetry.Trace.SpanK
-OpenTelemetry.Trace.TracerProvider.GetTracer(string! name, string? version = null) -> OpenTelemetry.Trace.Tracer!
+OpenTelemetry.Trace.TracerProvider.GetTracer(string! name) -> OpenTelemetry.Trace.Tracer!
+OpenTelemetry.Trace.TracerProvider.GetTracer(string! name, string? version = "") -> OpenTelemetry.Trace.Tracer!
+OpenTelemetry.Trace.TracerProvider.GetTracer(string! name, string? version = "", System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string!, object?>>? tags = null) -> OpenTelemetry.Trace.Tracer!
```
There are 2 things here:
1. It should go the Unshipped file.
2. It is breaking change, the contact should be `OpenTelemetry.Trace.TracerProvider.GetTracer(string! name, string? version = null, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string!, object?>>? tags = null) -> OpenTelemetry.Trace.Tracer!`

Kielek commented on src/OpenTelemetry.Api/Trace/TracerProvider.cs:
```diff
@@ -27,6 +27,18 @@ protected TracerProvider()
     /// </summary>
     public static TracerProvider Default { get; } = new TracerProvider();
 
+    /// <summary>
```
The public cotnact part IMO should looks like:

```csharp
    /// <summary>
    /// Gets a tracer with given name and version.
    /// </summary>
    /// <param name="name">Name identifying the instrumentation library.</param>
    /// <param name="version">Version of the instrumentation library.</param>
    /// <returns>Tracer instance.</returns>
    // 1.11.1 BACKCOMPAT OVERLOAD -- DO NOT TOUCH
    public Tracer GetTracer(
#if NET
        [AllowNull]
#endif
        string name,
        string? version) =>
        this.GetTracer(name, version);

    /// <summary>
    /// Gets a tracer with given name, version and tags.
    /// </summary>
    /// <param name="name">Name identifying the instrumentation library.</param>
    /// <param name="version">Version of the instrumentation library.</param>
    /// <param name="tags">Tags associated with the tracer.</param>
    /// <returns>Tracer instance.</returns>
    public Tracer GetTracer(
#if NET
        [AllowNull]
#endif
        string name,
        string? version = null,
        IEnumerable<KeyValuePair<string, object?>>? tags = null)
```

Ref: https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md

alimahboubi commented:
@Kielek Thanks for the detailed feedback! I've incorporated both of your suggestions:

- The new GetTracer overload is now in the Unshipped file and will be moved to Shipped during the release.

- I've updated the GetTracer signature to include the optional version and tags parameters, and adjusted the existing overload accordingly (removing the default parameter).

- I've updated the public contract part.

I've updated the commit with these changes. Please take another look!

Kielek commented on test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs:
```diff
@@ -408,6 +408,82 @@ static void InnerTest()
+        var tracer1 = tracerProvider.GetTracer("test", "1.0.0", tags1);
+        var tracer2 = tracerProvider.GetTracer("test", "1.0.0", tags2);
+
+        Assert.Equivalent(tracer1, tracer2);
```
```suggestion
        Assert.Same(tracer1, tracer2);
```

Kielek commented on test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs:
```diff
@@ -408,6 +408,82 @@ static void InnerTest()
+        var tracer1 = tracerProvider.GetTracer("test", "1.0.0", tags1);
+        var tracer2 = tracerProvider.GetTracer("test", "1.0.0", tags2);
+
+        Assert.Equivalent(tracer1, tracer2);
```
```suggestion
        Assert.Same(tracer1, tracer2);
```

Kielek commented on test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs:
```diff
@@ -408,6 +408,82 @@ static void InnerTest()
+        var tracerWithTags = tracerProvider.GetTracer("test", "1.0.0", tags);
+        var tracerWithoutTags = tracerProvider.GetTracer("test", "1.0.0");
+
+        Assert.NotSame(tracerWithTags, tracerWithoutTags);
```
```suggestion
        Assert.NotEqual(tracerWithTags, tracerWithoutTags);
```

Kielek commented on test/OpenTelemetry.Api.Tests/Trace/TracerTest.cs:
```diff
@@ -408,6 +408,82 @@ static void InnerTest()
+            // Activity started by the tracer with tags
+        }
+
+        Assert.Single(exportedItems);
+        var activity = exportedItems[0];
```
```suggestion
        var activity = Assert.Single(exportedItems);
```

Kielek reviewed (commented):
Your assertions are to weak. The goal is to have one instance per tags set. Now, you have multiple instances for the same set of tags.

If you include all changes, tests will be failing. You need to fix the TracerKey.

Kielek commented on src/OpenTelemetry.Api/Trace/TracerProvider.cs:
```diff
@@ -103,11 +118,94 @@ internal readonly record struct TracerKey
+        }
+
+        private static bool AreTagsEqual(
+            IEnumerable<KeyValuePair<string, object?>>? tags1,
+            IEnumerable<KeyValuePair<string, object?>>? tags2)
```
I think that you can use arrays here as parameters

Kielek commented on src/OpenTelemetry.Api/Trace/TracerProvider.cs:
```diff
@@ -103,11 +118,94 @@ internal readonly record struct TracerKey
+                return false;
+            }
+
+            return tags1.SequenceEqual(tags2);
```
And avoid Linq here, which are usually low-performant. consider for int i...

Kielek commented on src/OpenTelemetry.Api/Trace/TracerProvider.cs:
```diff
@@ -103,11 +118,94 @@ internal readonly record struct TracerKey
+                return null;
+            }
+
+            var orderedTagList = new List<KeyValuePair<string, object?>>(tags);
+            orderedTagList.Sort((left, right) => string.Compare(left.Key, right.Key, StringComparison.Ordinal));
```
Note for myself - inspired by ActivitySource code.

Kielek reviewed (commented):
Worth to add entry in CHANGELOG under OpenTelemetry.Api

xiang17 commented on src/OpenTelemetry.Api/Trace/TracerProvider.cs:
```diff
@@ -103,11 +118,121 @@ internal readonly record struct TracerKey
+            }
+
+            var orderedTagList = new List<KeyValuePair<string, object?>>(tags);
+            orderedTagList.Sort((left, right) => string.Compare(left.Key, right.Key, StringComparison.Ordinal));
```
Can there be multiple tags with the same key (and maybe with different value)? Does order matter for the tags with the same key?

alimahboubi replied:
Thanks for catching that! I checked the scenario with multiple tags that have the same key with different value and found an issue with it—so I went ahead and fixed it in the latest update.

xiang17 reviewed (approved):
Overall looks good.

rajkumar-rangaraj commented:
@alimahboubi Please resolve the conflicts.

alimahboubi commented:
> @alimahboubi Please resolve the conflicts.
I fixed it.
